pyiceberg/transforms.py

```diff
@@ -766,6 +858,47 @@ def _truncate_number(
         return None
 
 
+def _truncate_number_strict(
+    name: str, pred: BoundLiteralPredicate[L], transform: Callable[[Optional[L]], Optional[L]]
+) -> Optional[UnboundPredicate[Any]]:
+    boundary = pred.literal
+
+    if not isinstance(boundary, (LongLiteral, DecimalLiteral, DateLiteral, TimestampLiteral)):
+        raise ValueError(f"Expected a numeric literal, got: {type(boundary)}")
+
+    if isinstance(pred, BoundLessThan):
+        return LessThan(Reference(name), _transform_literal(transform, boundary))
+    elif isinstance(pred, BoundLessThanOrEqual):
+        return LessThan(Reference(name), _transform_literal(transform, boundary.increment()))  # type: ignore
+    elif isinstance(pred, BoundGreaterThan):
+        return GreaterThan(Reference(name), _transform_literal(transform, boundary))
+    elif isinstance(pred, BoundGreaterThanOrEqual):
+        return GreaterThan(Reference(name), _transform_literal(transform, boundary.decrement()))  # type: ignore
```

Comment:
Hi, I'm confused about decrement here. @Fokko 
E.g. col1 is date type and transform is month, for the predicate is `col1 >= 1970-01-02`, this will cause the predicate project to `month(col1) > -1`, e.g. the col1 is `1970-01-01`, `1970-01-01 >= 1970-01-02` is false but 
`month(1970-01-01) = 0 > -1` is true.

Reply:
Hmm, you might be onto something here: 

https://github.com/apache/iceberg-python/blob/88c4bad1d95175c3672bf7c47cba22ca803efcca/tests/test_transforms.py#L1006-L1009

https://github.com/apache/iceberg/blob/d402f83fc7b224b21242c506cf503e5bcbc8c867/api/src/test/java/org/apache/iceberg/transforms/TestDatesProjection.java#L139-L140

Luckily this code is not yet used, I'll write a patch tomorrow 👍

Reply:
Hey @ZENOTME Again, thanks for pointing this out. It turns out that most of these oddities are caused by a bug that was part of Iceberg `≤0.10.0`. We did not port this to PyIceberg: https://github.com/apache/iceberg/blob/ac6509a4e469f808bebb8b713a5c4213f98ff4a5/api/src/main/java/org/apache/iceberg/transforms/ProjectionUtil.java#L275

Reply:
Created a PR to refactor the tests: https://github.com/apache/iceberg-python/pull/1422